File: backend/app/schemas/inside_sales.py
"""Schemas for Inside Sales API."""

from datetime import datetime
from typing import Optional
from pydantic import Field
from app.schemas.base import CamelModel


class CallRecord(CamelModel):
    activity_id: str
    prospect_id: str
    agent_name: str
    agent_email: str
    event_code: int
    direction: str
    status: str
    call_start_time: str
    duration_seconds: int
    recording_url: str
    phone_number: str
    display_number: str
    call_notes: str
    call_session_id: str
    created_on: str
    last_eval_score: Optional[float] = None
    eval_count: int = 0


class CollectionFreshness(CamelModel):
    last_synced_at: datetime | None = None
    sync_in_progress: bool = False
    stale: bool = True


class CallListResponse(CamelModel):
    calls: list[CallRecord]
    total: int
    page: int
    page_size: int
    freshness: CollectionFreshness = Field(default_factory=CollectionFreshness)


class AgentListResponse(CamelModel):
    agents: list[str]


class LeadDetailResponse(CamelModel):
    prospect_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    cached: bool = False


class LeadListRecord(CamelModel):
    prospect_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    prospect_stage: str
    city: Optional[str] = None
    age_group: Optional[str] = None
    condition: Optional[str] = None
    hba1c_band: Optional[str] = None
    intent_to_pay: Optional[str] = None
    agent_name: Optional[str] = None
    rnr_count: int = 0
    answered_count: int = 0
    total_dials: int = 0
    connect_rate: Optional[float] = None      # None when total_dials == 0
    frt_seconds: Optional[int] = None         # None when null source or negative delta
    lead_age_days: int = 0
    days_since_last_contact: Optional[int] = None  # None when no activity yet
    mql_score: int = 0
    mql_signals: dict[str, bool] = Field(default_factory=dict)
    created_on: str
    last_activity_on: Optional[str] = None
    source: Optional[str] = None
    source_campaign: Optional[str] = None


class LeadListResponse(CamelModel):
    leads: list[LeadListRecord]
    total: int
    page: int
    page_size: int
    freshness: CollectionFreshness = Field(default_factory=CollectionFreshness)


class CollectionRefreshRequest(CamelModel):
    date_from: str | None = None
    date_to: str | None = None
    event_codes: str | None = None


class CollectionRefreshResponse(CamelModel):
    job_id: str
    source_family: str
    sync_mode: str
    status: str


class CollectionSyncStatus(CamelModel):
    """Durable freshness signal for a collection. Read from ``source_sync_runs``.

    ``lastSuccessAt`` is the most recent ``completed`` sync. ``lastAttemptAt``
    is the most recent attempt regardless of outcome. ``lastStatus`` /
    ``lastError`` describe that attempt so the UI can render failure state
    after a page reload (frontend cache is not durable across reloads).
    ``syncInProgress`` is true when any sync is currently ``running``.
    """
    last_success_at: datetime | None = None
    last_attempt_at: datetime | None = None
    last_status: str | None = None  # 'running' | 'completed' | 'failed' | 'cancelled'
    last_error: str | None = None
    sync_in_progress: bool = False


class LeadEvalHistoryEntry(CamelModel):
    """One evaluation record for a lead's call history."""
    id: str
    thread_id: str
    run_id: str
    result: dict                    # raw evaluator JSON
    created_at: str


class LeadCallRecord(CamelModel):
    activity_id: str
    call_time: str
    agent_name: Optional[str] = None
    duration_seconds: int = 0
    status: str
    recording_url: Optional[str] = None
    eval_score: Optional[float] = None
    is_counseling: bool = False              # duration_seconds >= 600


class LeadDetailFullResponse(CamelModel):
    # Profile
    prospect_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    prospect_stage: str
    city: Optional[str] = None
    age_group: Optional[str] = None
    condition: Optional[str] = None
    hba1c_band: Optional[str] = None
    blood_sugar_band: Optional[str] = None
    diabetes_duration: Optional[str] = None
    current_management: Optional[str] = None
    goal: Optional[str] = None
    intent_to_pay: Optional[str] = None
    job_title: Optional[str] = None
    preferred_call_time: Optional[str] = None
    agent_name: Optional[str] = None
    source: Optional[str] = None
    source_campaign: Optional[str] = None
    created_on: str
    # MQL
    mql_score: int = 0
    mql_signals: dict[str, bool] = Field(default_factory=dict)
    # Computed metrics
    frt_seconds: Optional[int] = None
    total_dials: int = 0
    connect_rate: Optional[float] = None
    counseling_count: int = 0
    counseling_rate: Optional[float] = None
    callback_adherence_seconds: Optional[int] = None
    lead_age_days: int = 0
    days_since_last_contact: Optional[int] = None
    # Call history
    call_history: list[LeadCallRecord] = Field(default_factory=list)
    history_truncated: bool = False
    # Eval history
    eval_history: list[LeadEvalHistoryEntry] = Field(default_factory=list)

File: src/services/api/insideSales.ts
import { apiRequest } from './client';

// ── Types ──────────────────────────────────────────────────────────────────

export interface LeadPlanPurchase {
  planName: string | null;
  durationOrQuantity: string | null;
  programPrice: string | null;
  invoiceAmount: string | null;
  paymentId: string | null;
  paymentDateAndTime: string | null;
  planAssignedAt: string | null;
  signUpDate: string | null;
  programStartDate: string | null;
  programEndDate: string | null;
  planIncludesCgm: string | null;
  cgm: string | null;
  cgmBrand: string | null;
  sensorCount: string | null;
  transmitterCount: string | null;
  bcaDevice: string | null;
  nutraceuticalsSold: string | null;
  salesTeam: string | null;
  deviceAwbNumber: string | null;
  leadConversionDate: string | null;
}

export interface LeadListRecord {
  prospectId: string;
  firstName: string;
  lastName: string | null;
  phone: string;
  prospectStage: string;
  city: string | null;
  ageGroup: string | null;
  condition: string | null;
  hba1cBand: string | null;
  intentToPay: string | null;
  agentName: string | null;
  rnrCount: number;
  answeredCount: number;
  totalDials: number;
  connectRate: number | null;
  frtSeconds: number | null;
  leadAgeDays: number;
  daysSinceLastContact: number | null;
  mqlScore: number;
  mqlSignals: Record<string, boolean>;
  createdOn: string;
  lastActivityOn: string | null;
  source: string | null;
  sourceCampaign: string | null;
  planName: string | null;
  plan: LeadPlanPurchase;
}

export interface LeadListResponse {
  leads: LeadListRecord[];
  total: number;
  page: number;
  pageSize: number;
  freshness: CollectionFreshness;
}

export interface CallRecord {
  activityId: string;
  prospectId: string;
  agentName: string;
  agentEmail: string;
  eventCode: number;
  direction: 'inbound' | 'outbound';
  status: string;
  callStartTime: string;
  durationSeconds: number;
  recordingUrl: string;
  phoneNumber: string;
  displayNumber: string;
  callNotes: string;
  callSessionId: string;
  createdOn: string;
  lastEvalScore?: number;
  evalCount?: number;
}

export interface CallListResponse {
  calls: CallRecord[];
  total: number;
  page: number;
  pageSize: number;
  freshness: CollectionFreshness;
}

export interface CollectionFreshness {
  lastSyncedAt: string | null;
  syncInProgress: boolean;
  stale: boolean;
}

export interface CallFilters {
  dateFrom: string;
  dateTo: string;
  agents: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  prospectId: string[];
  direction: string;
  status: string;
  hasRecording: boolean;
  eventCodes: string;
  durationMin: string;
  durationMax: string;
}

export interface LeadCallRecord {
  activityId: string;
  callTime: string;
  agentName: string | null;
  durationSeconds: number;
  status: string;
  recordingUrl: string | null;
  evalScore: number | null;
  isCounseling: boolean;
}

export interface LeadEvalHistoryEntry {
  id: string;
  threadId: string;
  runId: string;
  result: Record<string, unknown>;
  createdAt: string;
}

export interface LeadDetailFullResponse {
  prospectId: string;
  firstName: string;
  lastName: string | null;
  phone: string;
  email: string | null;
  prospectStage: string;
  city: string | null;
  ageGroup: string | null;
  condition: string | null;
  hba1cBand: string | null;
  bloodSugarBand: string | null;
  diabetesDuration: string | null;
  currentManagement: string | null;
  goal: string | null;
  intentToPay: string | null;
  jobTitle: string | null;
  preferredCallTime: string | null;
  agentName: string | null;
  source: string | null;
  sourceCampaign: string | null;
  createdOn: string;
  mqlScore: number;
  mqlSignals: Record<string, boolean>;
  frtSeconds: number | null;
  totalDials: number;
  connectRate: number | null;
  counselingCount: number;
  counselingRate: number | null;
  callbackAdherenceSeconds: number | null;
  leadAgeDays: number;
  daysSinceLastContact: number | null;
  callHistory: LeadCallRecord[];
  historyTruncated: boolean;
  evalHistory: LeadEvalHistoryEntry[];
  plan: LeadPlanPurchase;
}

// ── API functions ──────────────────────────────────────────────────────────

export interface LeadFilters {
  dateFrom: string;
  dateTo: string;
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  agents: string[];
  stage: string[];
  mqlMin: string;
  condition: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  city: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  prospectId: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  phone: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  planName: string[];
  q: string;
}

export type CallQueryScope = 'page' | 'all';
export type InsideSalesCollectionFamily = 'calls' | 'leads';

export interface CollectionRefreshResponse {
  jobId: string;
  sourceFamily: InsideSalesCollectionFamily;
  syncMode: string;
  status: string;
}

export interface CollectionCoverage {
  hasData: boolean;
  availableFrom: string | null;
  availableTo: string | null;
  lastScheduledSyncAt: string | null;
  lastScheduledSyncStatus: string | null;
}

export interface CollectionSyncStatus {
  lastSuccessAt: string | null;
  lastAttemptAt: string | null;
  lastStatus: 'running' | 'completed' | 'failed' | 'cancelled' | null;
  lastError: string | null;
  syncInProgress: boolean;
}

export async function fetchCollectionStatus(
  sourceFamily: InsideSalesCollectionFamily,
): Promise<CollectionSyncStatus> {
  return apiRequest<CollectionSyncStatus>(
    `/api/inside-sales/collections/${encodeURIComponent(sourceFamily)}/status`,
  );
}

export type SuggestionField =
  | 'prospect_id'
  | 'phone'
  | 'agent_name'
  | 'city'
  | 'stage'
  | 'plan_name';

export async function fetchCollectionSuggestions(
  sourceFamily: InsideSalesCollectionFamily,
  field: SuggestionField,
  q: string,
  limit = 20,
): Promise<string[]> {
  const params = new URLSearchParams({ field, limit: String(limit) });
  const trimmed = (q ?? '').trim();
  if (trimmed) params.set('q', trimmed);
  const res = await apiRequest<{ values: string[] }>(
    `/api/inside-sales/collections/${encodeURIComponent(sourceFamily)}/suggestions?${params.toString()}`,
  );
  return res.values ?? [];
}

export async function fetchCoverage(
  sourceFamily: InsideSalesCollectionFamily,
): Promise<CollectionCoverage> {
  return apiRequest<CollectionCoverage>(
    `/api/inside-sales/coverage?source_family=${encodeURIComponent(sourceFamily)}`,
  );
}

function parseCoverageDate(value: string | null | undefined): number | null {
  const trimmed = (value ?? '').trim();
  if (!trimmed) {
    return null;
  }
  const normalized = trimmed.includes('T') ? trimmed : `${trimmed.replace(' ', 'T')}Z`;
  const parsed = new Date(normalized).getTime();
  return Number.isNaN(parsed) ? null : parsed;
}

export function isRangeOutsideCoverage(
  coverage: CollectionCoverage | null,
  dateFrom: string,
  dateTo: string,
): boolean {
  const fromMs = parseCoverageDate(dateFrom);
  const toMs = parseCoverageDate(dateTo);
  const availableFromMs = parseCoverageDate(coverage?.availableFrom);
  const availableToMs = parseCoverageDate(coverage?.availableTo);

  if (
    !coverage
    || !coverage.hasData
    || fromMs === null
    || toMs === null
    || availableFromMs === null
    || availableToMs === null
  ) {
    return true;
  }

  return fromMs < availableFromMs || toMs > availableToMs;
}

export type CollectionRefreshSyncMode = 'incremental' | 'date_range' | 'bootstrap';

function buildCallSearchParams(
  filters: CallFilters,
  page: number,
  pageSize: number,
  scope: CallQueryScope,
): URLSearchParams {
  const params = new URLSearchParams({
    date_from: filters.dateFrom,
    date_to: filters.dateTo,
    page: String(page),
    page_size: String(pageSize),
  });

  if (scope !== 'page') {
    params.set('scope', scope);
  }
  if (filters.agents && filters.agents.length > 0) params.set('agents', filters.agents.join(','));
  if (filters.prospectId && filters.prospectId.length > 0) params.set('prospect_id', filters.prospectId.join(','));
  if (filters.direction) params.set('direction', filters.direction);
  if (filters.status) params.set('status', filters.status);
  if (filters.hasRecording) params.set('has_recording', 'true');
  if (filters.durationMin) params.set('duration_min', filters.durationMin);
  if (filters.durationMax) params.set('duration_max', filters.durationMax);
  if (filters.eventCodes) params.set('event_codes', filters.eventCodes);

  return params;
}

export async function fetchCalls(
  filters: CallFilters,
  page: number,
  pageSize: number,
  options?: { scope?: CallQueryScope },
): Promise<CallListResponse> {
  const scope = options?.scope ?? 'page';
  const params = buildCallSearchParams(filters, page, pageSize, scope);
  return apiRequest<CallListResponse>(`/api/inside-sales/calls?${params.toString()}`);
}

export async function fetchCallsForSelection(
  filters: CallFilters,
  pageSize = 500,
): Promise<CallListResponse> {
  return fetchCalls(filters, 1, pageSize, { scope: 'all' });
}

export async function fetchLeads(
  filters: LeadFilters,
  page: number,
  pageSize: number,
): Promise<LeadListResponse> {
  const params = new URLSearchParams({
    date_from: filters.dateFrom,
    date_to: filters.dateTo,
    page: String(page),
    page_size: String(pageSize),
  });
  const q = (filters.q ?? '').trim();
  if (filters.agents && filters.agents.length > 0) params.set('agents', filters.agents.join(','));
  if (filters.stage && filters.stage.length > 0) params.set('stage', filters.stage.join(','));
  if (filters.mqlMin) params.set('mql_min', filters.mqlMin);
  if (filters.condition && filters.condition.length > 0) params.set('condition', filters.condition.join(','));
  if (filters.city && filters.city.length > 0) params.set('city', filters.city.join(','));
  if (filters.prospectId && filters.prospectId.length > 0) params.set('prospect_id', filters.prospectId.join(','));
  if (filters.phone && filters.phone.length > 0) params.set('phone', filters.phone.join(','));
  if (filters.planName && filters.planName.length > 0) params.set('plan_name', filters.planName.join(','));
  if (q) params.set('q', q);

  return apiRequest<LeadListResponse>(`/api/inside-sales/leads?${params.toString()}`);
}

export async function fetchLeadDetail(
  prospectId: string,
  options?: { refresh?: boolean },
): Promise<LeadDetailFullResponse> {
  const qs = options?.refresh ? '?refresh=true' : '';
  return apiRequest<LeadDetailFullResponse>(`/api/inside-sales/leads/${prospectId}/detail${qs}`);
}

export async function refreshInsideSalesCollection(
  sourceFamily: InsideSalesCollectionFamily,
  payload: {
    syncMode?: CollectionRefreshSyncMode;
    dateFrom?: string;
    dateTo?: string;
    eventCodes?: string;
    overlapMinutes?: number;
  },
): Promise<CollectionRefreshResponse> {
  return apiRequest<CollectionRefreshResponse>(`/api/inside-sales/collections/${sourceFamily}/refresh`, {
    method: 'POST',
    body: JSON.stringify({
      syncMode: payload.syncMode,
      dateFrom: payload.dateFrom,
      dateTo: payload.dateTo,
      eventCodes: payload.eventCodes,
      overlapMinutes: payload.overlapMinutes,
    }),
  });
}
